examples/server/api_like_OAI.py

```diff
@@ -46,18 +50,184 @@ def convert_chat(messages):
+            elemtype = schema_typename(prefix, schema['items'], defs, arrs)
+            typename = f'array-{elemtype}'
+            if not is_basetype(elemtype):
+                typename = prefix + typename
+            typename = format_rulename(typename)
+            if typename not in arrs:
+                arrs[typename] = elemtype
+        return typename
+
+    def arr_to_rules(rules, prefix, name, elemtype):
+        rulename = name
+        rulename = format_rulename(rulename)
+        rules[rulename] = f'"[" ( {elemtype} ( "," {elemtype} )* )? "]"'
+
+    def enum_to_rules(rules, prefix, name, schema):
+        enum_values = schema['enum']
+        etype = schema['type']
+        def value_pattern(value):
+            if etype == 'string':
+                return f'"\\"{repr(value)[1:-1]}\\""'
+            else:
+                return repr(value)
+        values = ' | '.join([
+            value_pattern(value)
+            for value in enum_values
+        ])
+        rulename = format_rulename(f'{prefix}{name}')
+        rules[rulename] = f'( {values} )'
+
+    def anyof_to_rules(rules, prefix, name, schema, defs, arrs):
+        values = schema['anyOf']
+        values = ' | '.join([
+            schema_typename(prefix, value, defs, arrs)
+            for value in values
+        ])
+        rulename = format_rulename(f'{prefix}{name}')
+        rules[rulename] = f'( {values} )'
+
+    def declare_rules(indent, rules, prefix, name, schema, defs, arrs):
+        if 'enum' in schema:
+            enum_to_rules(rules, prefix, format_rulename(name), schema)
+        elif 'anyOf' in schema:
+            anyof_to_rules(rules, prefix, format_rulename(name), schema, defs, arrs)
+        elif schema.get('type', None) == 'object':
+            obj_to_rules(indent + indent_inc, rules, prefix, format_rulename(name), schema, defs, arrs, is_toplevel=False)
+
+    def obj_to_rules(indent, rules, prefix, name, schema, defs, arrs, is_toplevel):
+        assert(schema['type'] == 'object')
+        if defs is None:
+            defs = {}
+        if arrs is None:
+            arrs = {}
+
+
+        rulename = name
+        if not is_toplevel:
+            rulename = prefix+rulename
+        rulename = format_rulename(rulename)
+
+        if '$defs' in schema:
+            for name, _def in schema['$defs'].items():
+                defs['#/$defs/' + name] = format_rulename(prefix + name)
+
+        if '$defs' in schema:
+            for name, _def in schema['$defs'].items():
+                declare_rules(indent + indent_inc, rules, prefix, name, _def, defs, arrs)
+
+        for name, prop in schema['properties'].items():
+            declare_rules(indent + indent_inc, rules, prefix, name, prop, defs, arrs)
+
+        def propery_to_grammar(name, typename):
+            return f'"\\"" "{name}" "\\"" ":" {typename}'
+
+        properties = ' "," '.join([
+            propery_to_grammar(name, schema_typename(prefix, prop, defs, arrs))
```

Comment:
Running `make_grammar` in isolation, I'm noticing nested objects get rendered as `{func}-object` rather than the specific rule name. I believe it comes down to the use of `schema_typename` here?
```
>>> print(make_grammar([{'name': 'foo', 'parameters': {'type': 'object', 'properties': {'a': {'type': 'object', 'properties': {'b': {'type': 'object', 'properties': {'c': {'type': 'number'}}}}}}}}], {'name': 'foo'}))
foo-b ::= "{" "\"" "c" "\"" ":" number "}"
foo-a ::= "{" "\"" "b" "\"" ":" foo-object "}"
foo ::= "{" "\"" "a" "\"" ":" foo-object "}"
```